Occasionally failing Python tests caused by caching

Occasionally, one of the Python tests aborts because of an invalid value in the verification cache. This is probably a data race caused by the parallel execution of multiple tests.

```
tests/unit/cli_test.py:64: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
rflx/cli.py:90: in main
    args.func(args)
rflx/cli.py:111: in generate
    model = parse(args.files)
rflx/cli.py:124: in parse
    parser = Parser(skip_verification, cached=True)
rflx/specification/parser.py:62: in __init__
    self.__cache = Cache(cached)
rflx/specification/cache.py:23: in __init__
    self._load_cache()
rflx/specification/cache.py:54: in _load_cache
    self._verification = json.load(f)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/json/__init__.py:296: in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/json/__init__.py:348: in loads
    return _default_decoder.decode(s)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/json/decoder.py:337: in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <json.decoder.JSONDecoder object at 0x7feb6e67e710>, s = '', idx = 0

    def raw_decode(self, s, idx=0):
        """Decode a JSON document from ``s`` (a ``str`` beginning with
        a JSON document) and return a 2-tuple of the Python
        representation and the index in ``s`` where the document ended.
    
        This can be used to decode a JSON document from a string that may
        have extraneous data at the end.
    
        """
        try:
            obj, end = self.scan_once(s, idx)
        except StopIteration as err:
>           raise JSONDecodeError("Expecting value", s, err.value) from None
E           json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/json/decoder.py:355: JSONDecodeError
=========================== short test summary info ============================
FAILED tests/unit/cli_test.py::test_main_generate - json.decoder.JSONDecodeEr...
```

https://github.com/Componolit/RecordFlux/runs/1678871112?check_suite_focus=true

--- a/rflx/specification/cache.py
+++ b/rflx/specification/cache.py
@@ -19,7 +19,6 @@ class Cache:
 
         self._verification: Dict[str, str] = {}
 
-        self._initialize_cache()
         self._load_cache()
 
     def is_verified(self, message: AbstractMessage) -> bool:
@@ -42,18 +41,15 @@ class Cache:
             self._verification[message.full_name] = message_hash
             self._write_cache()
 
-    @staticmethod
-    def _initialize_cache() -> None:
-        CACHE_DIR.mkdir(parents=True, exist_ok=True)
-        if not (CACHE_DIR / VERIFICATION_FILE).exists():
-            with open(CACHE_DIR / VERIFICATION_FILE, "w") as f:
-                json.dump({}, f)
-
     def _load_cache(self) -> None:
-        with open(CACHE_DIR / VERIFICATION_FILE) as f:
-            self._verification = json.load(f)
+        try:
+            with open(CACHE_DIR / VERIFICATION_FILE) as f:
+                self._verification = json.load(f)
+        except (FileNotFoundError, json.JSONDecodeError):
+            pass
 
     def _write_cache(self) -> None:
+        CACHE_DIR.mkdir(parents=True, exist_ok=True)
         with open(CACHE_DIR / VERIFICATION_FILE, "w") as f:
             json.dump(self._verification, f)
 

--- a/rflx/specification/cache.py
+++ b/rflx/specification/cache.py
@@ -59,4 +59,5 @@ class Cache:
 
     @staticmethod
     def _message_hash(message: AbstractMessage) -> str:
-        return hashlib.md5(f"{__version__}|{message!r}".encode("utf-8")).hexdigest()
+        types = "|".join(str(t) for t in message.types.values())
+        return hashlib.md5(f"{__version__}|{message}|{types}".encode("utf-8")).hexdigest()

--- a/rflx/specification/parser.py
+++ b/rflx/specification/parser.py
@@ -1201,7 +1201,7 @@ class Parser:
             *model.INTERNAL_TYPES.values(),
         ]
         self.__sessions: List[model.Session] = []
-        self.__cache = Cache(cached)
+        self.__cache = Cache(not skip_verification and cached)
 
     def __convert_unit(
         self, spec: Specification, filename: Path, transitions: List[ID] = None

--- a/tests/unit/specification/cache_test.py
+++ b/tests/unit/specification/cache_test.py
@@ -8,24 +8,21 @@ from tests.data.models import TLV_MESSAGE
 def test_init(tmp_path: Path) -> None:
     cache.CACHE_DIR = tmp_path / "Test"
     cache.Cache()
-    assert (tmp_path / "Test").is_dir()
-    assert (tmp_path / "Test" / cache.VERIFICATION_FILE).is_file()
+    assert not (tmp_path / "Test").exists()
 
 
-def test_init_existing(tmp_path: Path) -> None:
+def test_init_valid(tmp_path: Path) -> None:
     cache.CACHE_DIR = tmp_path
     with open(tmp_path / cache.VERIFICATION_FILE, "x") as f:
         f.write("{}")
     cache.Cache()
-    assert tmp_path.is_dir()
-    assert (tmp_path / cache.VERIFICATION_FILE).is_file()
 
 
-def test_init_disabled(tmp_path: Path) -> None:
-    cache.CACHE_DIR = tmp_path / "Test"
-    cache.Cache(enabled=False)
-    assert not (tmp_path / "Test").is_dir()
-    assert not (tmp_path / "Test" / cache.VERIFICATION_FILE).is_file()
+def test_init_invalid(tmp_path: Path) -> None:
+    cache.CACHE_DIR = tmp_path
+    with open(tmp_path / cache.VERIFICATION_FILE, "x") as f:
+        f.write("invalid")
+    cache.Cache()
 
 
 def test_verified(tmp_path: Path) -> None:

